src/frontend/src/optimizer/plan_node/generic/scan.rs

```diff
@@ -37,8 +39,8 @@ pub struct Scan {
     /// Include `output_col_idx` and columns required in `predicate`
     pub required_col_idx: Vec<usize>,
     pub output_col_idx: Vec<usize>,
-    /// Descriptor of the table
-    pub table_desc: Rc<TableDesc>,
+    /// Table Catalog of the upstream table that the descriptor is derived from.
+    pub table_catalog: Arc<TableCatalog>,
```

Comment:
Which fields used in the arrangement backfill upstream table exist only in `TableCatalog ` but not in `TableDesc`?

Reply:
We use the function `from_table_catalog` to construct state table. Let me see if it's possible to derive it from `TableDesc`.

Reply:
related https://github.com/risingwavelabs/risingwave/issues/13146#issuecomment-1797674824

Reply:
> Which fields used in the arrangement backfill upstream table exist only in TableCatalog but not in TableDesc?

- table option (can technically be passed in to table desc, since now table desc contains the sole field (retention seconds))
- vnode_col_index, just a single field, I suppose we can also include it in table desc as well.

So I think we could use `TableDesc` too.

The rest of the fields can be derived in some way, although they may not be present in `TableDesc`.

Is there some reason to prefer `TableDesc` vs `TableCatalog`? 🤔 My thought process is that it's the least error prone to just clone the upstream catalog.

Is it just to reduce the refactoring changes to frontend?

Reply:
Hmm that will require using and modifying the proto definition of `StorageTableDesc`, which does not seem preferred... That should be reserved for `StorageTable`.

Reply:
I will open a separate PR to do some refactoring. Seems like `TableCatalog` can be used everywhere in frontend instead, so we shouldn't need to maintain a separate interface, just for `TableDesc`.

Reply:
Separated `arrangement backfill`'s proto definition from the stream proto definition.

We can freely change it ourselves, it won't be exposed to users.

Reply:
Still use `TableCatalog` (via `catalog.Table`) to construct replicated state table.

In the future we can do further refactoring: https://github.com/risingwavelabs/risingwave/issues/13868

Reply:
Seems like it will cause a lot of breakage, since several areas of code expect `StreamScan`. I will parameterize the upstream table instead using `oneof`.